step-controller/step-controller-server/src/main/java/step/core/controller/ControllerSettingAccessorImpl.java

```diff
@@ -18,20 +18,136 @@
  ******************************************************************************/
 package step.core.controller;
 
+import org.bson.types.ObjectId;
 import step.core.accessors.AbstractAccessor;
 import step.core.collections.Collection;
 import step.core.collections.Filters;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
 public class ControllerSettingAccessorImpl extends AbstractAccessor<ControllerSetting> implements ControllerSettingAccessor {
-	
+
+	private final Map<String, List<ControllerSettingHook>> hooksMap = new ConcurrentHashMap<>();
+
 	public ControllerSettingAccessorImpl(Collection<ControllerSetting> collectionDriver) {
 		super(collectionDriver);
 	}
 
 	public ControllerSetting getSettingByKey(String key) {
 		return collectionDriver.find(Filters.equals("key", key), null, null, null, 0).findFirst().orElse(null);
 	}
-	
+
+	@Override
+	public void addHook(String key, ControllerSettingHook hook) {
+		this.hooksMap.computeIfAbsent(key, k -> new ArrayList<>());
+		List<ControllerSettingHook> list = this.hooksMap.get(key);
+		list.add(hook);
+	}
+
+	@Override
+	public boolean removeHook(String key, ControllerSettingHook hook) {
+		List<ControllerSettingHook> hooks = getHooksBySettingKey(key);
+		if(hooks != null){
+			return hooks.remove(hook);
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public ControllerSetting save(ControllerSetting entity) {
+		// we can change the key of existing setting - in this case we notify hooks about deleted/created setting
+		ControllerSetting oldValueWithChangedKey = getOldValueWithAnotherKey(entity);
+		ControllerSetting res = super.save(entity);
+
+		if (oldValueWithChangedKey != null) {
+			List<ControllerSettingHook> hooksOnDelete = getHooksBySettingKey(oldValueWithChangedKey.getKey());
+			if (hooksOnDelete != null) {
+				for (ControllerSettingHook hook : hooksOnDelete) {
+					hook.onSettingRemove(oldValueWithChangedKey.getId(), oldValueWithChangedKey);

```

Comment:
@iegorov777 The code in lines 67 - 81 seems to be duplicated in lines 112..., for readiness and maintenance, please extract it in a dedicated method